Initial proxy

NathanSegerlind opened the pull request:
Initial implementation of the proxy pipeline off the dev branch.

The model is not yet tuned or rigorously evaluated, but it scores a day of data within minutes and the scores demonstrate a "spread" in values.

Does not yet incorporate feedback from OA.

NathanSegerlind committed:
initial_proxy

initial work on initial (untuned!!!) implementation of proxy threat model

NathanSegerlind committed:
initial_proxy

Initial proxy implementation.
It builds.

NathanSegerlind committed:
initial_proxy

First implementation of proxy pipeline... not tuned or evaluated, but it does score proxy data.

Requires that PROXY_PATH be defined in duxbay.conf

NathanSegerlind commented:
this change requires that PROXY_PATH be defined in duxbay.conf  (see 
https://github.com/Open-Network-Insight/oni-setup/pull/17 )

NathanSegerlind committed:
initial_proxy

Initial implementation of the proxy pipeline.

Feedback not yet supported.

Model not tuned or evaluated.

Does score a day of data very quickly and provide a 'spead' of scores.

rabarona commented on src/main/scala/org/opennetworkinsight/ProxyPostLDA.scala:
```diff
@@ -0,0 +1,116 @@
+
+    var df_cols = new Array[String](0)
+
+    val multidata = {
```
We should probably get ride of this multidata thing.

NathanSegerlind replied:
not a bad suggestion... I will take it out when I port to the new architecture

rabarona commented on src/main/scala/org/opennetworkinsight/ProxyPostLDA.scala:
```diff
@@ -0,0 +1,116 @@
+
+    logger.info("Persisting data")
+    val scored = src_scored.sortByKey().map(row => row._2.mkString(","))
+    scored.persist(StorageLevel.MEMORY_AND_DISK)
```
No real need to persist and then save to TextFile. You can just saveAsText.

rabarona commented:
Overall it looks good, there are changes that are going to be required to match new architecture but I'm sure you are already aware of them.

LGTM.

NathanSegerlind committed:
initial_proxy

Updated output fields to provide data required by the OA component.

LedaLima commented:
I'm still missing a few fields:

- uriport          
- uripath          
- uriquery          
- serverip         
- scbytes          
- csbytes           
- subdomainentropy
- top domain

Also, we had proposed to remove the 'proxy_' prefix from all the field names (my bad, I'll update the spreadsheet) to keep uniformity with the other pipelines. At least the proxy_scores.csv file will not contain that prefix.

NathanSegerlind committed:
initial_proxy

added some more fields to the output

NathanSegerlind committed:
Merge branch 'dev' into initial_proxy

LedaLima commented:
LGTM !

NathanSegerlind committed:
initial_proxy

Got the initial proxy pipeline working on the new ML architecture.

Expanded command line arguments in ml_ops.sh to their verbose forms.

NathanSegerlind committed:
initial_proxy

Comment edit in ml_ops.sh

NathanSegerlind commented:
I've moved things to the updated architecture.

rabarona commented on ml_ops.sh:
```diff
@@ -72,30 +74,30 @@ hdfs dfs -rm -R -f ${HDFS_SCORED_CONNECTS}
+  --nodes ${nodes} \
+  --scored ${HDFS_SCORED_CONNECTS} \
+  --threshold ${TOL}
 
```
I see you did the change, I had it done in another branch too.

rabarona commented on src/main/scala/org/opennetworkinsight/LDA.scala:
```diff
@@ -23,21 +22,19 @@ object LDA {
-        apacheLogger.getLogger("akka").setLevel(Level.OFF)
+        Logger.getLogger("org").setLevel(Level.OFF)
+        Logger.getLogger("akka").setLevel(Level.OFF)
 
```
The idea was to differentiate between the logger variable and Logger class. I guess it looks good anyways.

rabarona commented on src/main/scala/org/opennetworkinsight/LDAWrapper.scala:
```diff
@@ -24,6 +24,7 @@ object LDAWrapper {
   =
   {
 
+    println("Proxy topic analysis: gathering words")
```
I agree with adding descriptive log like "Topic analysis: gathering words" but I don't think we should use Proxy or any data source name in this file.

NathanSegerlind replied:
that wasn't supposed to be checked in... cleaning it out now

rabarona commented on src/main/scala/org/opennetworkinsight/ProxyLDA.scala:
```diff
@@ -0,0 +1,31 @@
+    val docWordCount = ProxyPreLDA.proxyPreLDA(config.inputPath, config.scoresFile, config.duplicationFactor, sparkContext,
+      sqlContext, logger)
+
+    val ldaResult = LDAWrapper.runLDA(docWordCount, config.modelFile, config.topicDocumentFile, config.topicWordFile,
```
I just learned how to return a tuple instead of having to create a Map[String, RDD[String]].
That will change val ldaResult to val (documentResults, wordResults) = LDAWrapper(....)

Then in the next line (25) we can use just documentResults instead of ldaResults("document_results").

Will keep you posted.

NathanSegerlind replied:
sounds good... but can come in a future checkin

rabarona commented on src/main/scala/org/opennetworkinsight/ProxyPostLDA.scala:
```diff
@@ -0,0 +1,126 @@
+
+    val logger = LoggerFactory.getLogger(this.getClass)
+    ApacheLogger.getLogger("org").setLevel(Level.OFF)
+    ApacheLogger.getLogger("akka").setLevel(Level.OFF)
```
I don't think we need this code - either. Logger is being passed to proxyPostLDA. 
Also apacheLogger doesn't need to be re-configured/turned off

NathanSegerlind replied:
doh! cut & paste gets you every time

natedogs911 commented:
LGTM!!!!

rabarona commented on src/main/scala/org/opennetworkinsight/ProxyPreLDA.scala:
```diff
@@ -0,0 +1,100 @@
+
+    val logger = LoggerFactory.getLogger(this.getClass)
+    apacheLogger.getLogger("org").setLevel(Level.OFF)
+    apacheLogger.getLogger("akka").setLevel(Level.OFF)
```
I don't think we need this code - either. Logger is being passed to proxyPostLDA. 
Also apacheLogger doesn't need to be re-configured/turned off

rabarona commented on src/main/scala/org/opennetworkinsight/ProxyPostLDA.scala:
```diff
@@ -0,0 +1,126 @@
+          df = df.unionAll(sqlContext.parquetFile(file).filter("proxy_date is not null and proxy_time is not null and proxy_clientip is not null"))
+        }
+      }
+      df = df.select("proxy_date",
```
I see what you did. For pre you are just pulling the necessary columns for word formation and LDA, then in post you are getting all the required columns for OA.

💯  👍

NathanSegerlind replied:
I don't like how "scoring" is pulling in columns for OA, but that's how it is right now... the trick is joining the scored records with the values in the other columns... I don't think there's a good unique key so it might be tricky to separate scoring from pulling data for OA (even though that's how I think it should be)

just something to think about

rabarona replied:
Agree. The biggest challenge is doing "join" with big amounts of data.

NathanSegerlind committed:
initial_proxy

clean up, removal of test code

NathanSegerlind commented:
okay, I've addressed Ricardo's comments

rabarona commented on src/main/scala/org/opennetworkinsight/ProxyPostLDA.scala:
```diff
@@ -0,0 +1,126 @@
+    logger.info("Persisting data")
+
+    val scored = src_scored.filter({case (score, row) => score < threshold}).sortByKey().map({case (score, row) => row.mkString(",")})
+    scored.persist(StorageLevel.MEMORY_AND_DISK)
```
No need to persist in memory and disk and then save. We can just go straight to saveAsTextFile.

NathanSegerlind committed:
initial_proxy

eliminated unnecessary persistence of data

rabarona commented:
LGTM

natedogs911 commented:
LGTM again
